### PR TITLE
[RFC] Enable syntax highlighting and filetype stuff by default

### PR DESCRIPTION
**Pull request by fmoralesc**
Re: https://github.com/neovim/neovim/issues/1664

This behaviour is switched off with `-u NONE`, so the tests didn't need to be changed.

### CODE REVIEW

**Comment by splinterofchaos** [comment]
> This behaviour is switched off with -u NONE

I don't think that's a good idea. `-u NONE` creates a baseline for diagnosing bugs and should put the editor in exactly the same state as loading an empty `vimrc`. Maybe I'm being a little paranoid, but perhaps there's another way.

**Comment by fmoralesc** [comment]
The changes to the tests are minimal, so I could just disable that check. The problem with that is the loading of these files in the runtime pollutes somewhat the global namespace of variables, so some stuff needed to be added to the tests that doesn't really make much sense there.

> -u NONE creates a baseline for diagnosing bugs and should put the editor in exactly the same state as loading an empty vimrc

I feel `-u NORC` is meant for that case, and `-u NONE` for bypassing loading stuff from the runtime paths. I think `NONE` is used for the tests because it provides the minimal baseline. But I know the convention so far is not that.

**Comment by fmoralesc** [comment]
I'm marking this as RFC, because there is not much else to include. 

I think the `-u NONE` behavior is safe because it results in nvim behaving exactly as it does now, which is not to enable syntax highighting or filetype plugins, so the "testing baseline" isn't changed. To disable user configuration, there's always `-u NORC`, and it's just a matter of policy to point the user to use that when reporting bugs that might be related to the runtime.

**Comment by pvinis** [comment]
any news on this?

**Comment by justinmk** [comment]
Targetting 0.2 milestone, this may require some thought.

**Comment by fmoralesc** [comment]
@justinmk :+1: Good call. 

This is related to the handling of plugin managers, so those things might need to be taken on together.

**Comment by pvinis** [comment]
cool. maybe a silly question, but isnt syntax on already? is it because i have an rc file?

**Comment by Pyrohh** [comment]
@pvinis do you use vim-plug? IIRC it does `filetype plugin indent on` and `syntax enable`?

**Comment by pvinis** [comment]
yes. then thats why

On Παρ, Ιουλ 24, 2015 at 20:03, neovim/neovim
<reply@reply.github.com>
wrote:
@pvinis [https://github.com/pvinis] do you use vim-plug? IIRC it does filetype plugin indent on and syntax enable ?

—
Reply to this email directly or view it on GitHub
[https://github.com/neovim/neovim/pull/2675#issuecomment-124600725] .[https://github.com/notifications/beacon/AAGHiaX1P-HQW5m58TY-jTcI1Kdsb0rdks5ognVfgaJpZM4Ec3A2.gif]

**Comment by fmoralesc** [comment]
Anyway, I rebased this to master. Travis failed on the OSX gcc-4.9 build, but it seems unrelated.

File: src/nvim/main.c [hunk 1]
```diff
@@ -1439,6 +1445,28 @@ static void load_plugins(void)
...
+ * Enable syntax highlighting.
+ */
+static void enable_syntax_highlighting(mparm_T *parmp)
+{
+  if (parmp->use_vimrc != NULL && strcmp(parmp->use_vimrc, "NONE") == 0) {
+      return;
+  }
+  (void)do_source((char_u *)SYNTAX_FILE, FALSE, DOSO_NONE);
+}
+
```
Comment (Pyrohh) [inline src/nvim/main.c]: Instead of all this, why not just do:
```c

// Don't do if `-u NONE` was passed on the command line
if (params->use_vimrc != NULL && strcmp(params->use_vimrc, "NONE") != 0) {
  // :filetype plugin indent enable
  (void)force_enable_filetype();
  // enable syntax highlighting
  (void)do_source((char_u *)SYNTAX_FILE, FALSE, DOSO_NONE);
}

```

File: runtime/doc/vim_diff.txt [hunk 2]
```diff
@@ -28,6 +28,10 @@ these differences.
 ==============================================================================
 2. Option defaults					  *nvim-option-defaults*
```
Comment (Pyrohh) [inline runtime/doc/vim_diff.txt]: Given this addition, I think the section should be renamed to just `Defaults`.

File: runtime/doc/vim_diff.txt [hunk 3]
```diff
@@ -28,6 +28,10 @@ these differences.
...
 
+- syntax highlighting is enabled by default (can be disabled with the "-u NONE"
+  cmdline argument)
+- filetype-related plugins and scripts are enabled by default (can be disabled
+  with the "-u NONE" cmdline argument)
```
Comment (Pyrohh) [inline runtime/doc/vim_diff.txt]: perhaps add a newline between these top two and everything below? Also, why not just say `plugins' instead of `plugins and scripts'? They sound the same to me.

Comment (Pyrohh) [inline runtime/doc/vim_diff.txt]: oh, and I think Vim already has a concept of a command line (`:`), but I wonder what other wording could be used.

**Comment by Pyrohh** [comment]
I think it'd be a good idea to make more prominent notices at the top of the tests you modified.

File: src/nvim/main.c (hunk 1 above)
Comment (fmoralesc) [inline src/nvim/main.c]: Just because I put these in separate commits, I guess. I would have to ask old me.

File: runtime/doc/vim_diff.txt (hunk 2 above)
Comment (fmoralesc) [inline runtime/doc/vim_diff.txt]: :+1:

File: runtime/doc/vim_diff.txt (hunk 3 above)
Comment (fmoralesc) [inline runtime/doc/vim_diff.txt]: I put "plugins and scripts" because this loads the filetype and ftplugin files plus the indent file, and in my mind only the second(ftplugin) defines proper "plugins".

**Comment by fmoralesc** [comment]
> I think it'd be a good idea to make more prominent notices at the top of the tests you modified.

Sure.

File: runtime/doc/vim_diff.txt (hunk 3 above)
Comment (Pyrohh) [inline runtime/doc/vim_diff.txt]: :+1:

**Comment by jleclanche** [comment]
I'm +1 on adding `syntax on` by default, it's very certainly what the majority of vim/nvim instances want.

**Comment by fmoralesc** [comment]
Ok, then. Are we going ahead with this, @justinmk? If so, please (all of you who might have some interest in this) take a new look at the PR so we can merge this soon.

File: runtime/doc/starting.txt [hunk 4]
```diff
@@ -445,31 +447,43 @@ accordingly.  Vim proceeds in this order:
...
 
-8. Perform GUI initializations
+10. Perform GUI initializations
 	Only when starting "gvim", the GUI initializations will be done.  See
 	|gui-init|.
 
-9. Read the ShaDa file
-	If the 'shada' option is not empty, the ShaDa file is read.  See
+11. Read the ShaDa file
+	If the 'shada' option is not empty, the viminfo file is read.  See
```
Comment (Pyrohh) [inline runtime/doc/starting.txt]: I guess this line shouldn't be here?

File: runtime/doc/vim_diff.txt [hunk 5]
```diff
@@ -28,7 +28,11 @@ these differences.
   session information.
 
 ==============================================================================
-2. Option defaults					  *nvim-option-defaults*
+2. Defaults					            *nvim-defaults*
+
+- Syntax highlighting is enabled by default
+- Filetype-related plugins and scripts are enabled by default
+  Note: these defaults can be disabled with the "-u NONE" cmdline argument
```
Comment (Pyrohh) [inline runtime/doc/vim_diff.txt]: Maybe `cmdline` -> `command line`, since `:h cmdline` is regarding ex commands.

A `|-u|` link a at the end might be useful too, like in `starting.txt:401`

File: src/nvim/main.c [hunk 6]
```diff
@@ -341,6 +341,17 @@ int main(int argc, char **argv)
...
 
   /*
+   * If using the runtime (-u is not NONE),
+   * enable syntax and filetype plugins
+   */
+  if (params.use_vimrc != NULL && strcmp(params.use_vimrc, "NONE") != 0) {
+    // Enable syntax highlighting.
+    (void)do_source((char_u *)SYNTAX_FILE, false, DOSO_NONE);
+    // :filetype plugin indent enable
+    (void)force_enable_filetype();
```
Comment (Pyrohh) [inline src/nvim/main.c]: this function returns void, so no need for the cast

File: runtime/doc/starting.txt (hunk 4 above)
Comment (fmoralesc) [inline runtime/doc/starting.txt]: :+1:

File: runtime/doc/vim_diff.txt (hunk 5 above)
Comment (fmoralesc) [inline runtime/doc/vim_diff.txt]: :+1:

File: src/nvim/os/unix_defs.h [hunk 7]
```diff
@@ -28,6 +28,9 @@
 #ifndef SYNTAX_FNAME
 # define SYNTAX_FNAME   "$VIMRUNTIME/syntax/%s.vim"
 #endif
+#ifndef SYNTAX_FILE
+# define SYNTAX_FILE   "$VIMRUNTIME/syntax/syntax.vim"
+#endif
```
Comment (Pyrohh) [inline src/nvim/os/unix_defs.h]: should probably add a stub in `os/win_defs.h` too

Comment (fmoralesc) [inline src/nvim/os/unix_defs.h]: Should it? None of the other items are present there.

Comment (Pyrohh) [inline src/nvim/os/unix_defs.h]: Not sure, it's not a big deal for now IMO.

**Comment by mhinz** [comment]
LGTM!

**Comment by fmoralesc** [comment]
Thanks, @mhinz. I'll rebase and wait if anyone else has more comments.

File: runtime/doc/starting.txt [hunk 8]
```diff
@@ -445,31 +447,42 @@ accordingly.  Vim proceeds in this order:
 	commands from the command line have not been executed yet.  You can
 	use "--cmd 'set noloadplugins'" |--cmd|.
 
-5. Set 'shellpipe' and 'shellredir'
+5. Enable filetype plugins and indent scripts.
```
Comment (justinmk) [inline runtime/doc/starting.txt]: Enable filetype and indent plugins.

File: runtime/doc/vim_diff.txt [hunk 9]
```diff
@@ -11,7 +11,7 @@ the "{Nvim}" tag.  This document is a complete and centralized list of all
 these differences.
 
 1. Configuration		|nvim-configuration|
-2. Option defaults		|nvim-option-defaults|
+2. Defaults			|nvim-defaults|
```
Comment (justinmk) [inline runtime/doc/vim_diff.txt]: :+1:

File: src/nvim/os/unix_defs.h (hunk 7 above)
Comment (justinmk) [inline src/nvim/os/unix_defs.h]: Should be able to move to globals.h and use `_PATH_SEP`

Comment (justinmk) [inline src/nvim/os/unix_defs.h]: Actually I thought this was already done, when last was this rebased?

File: src/nvim/main.c [hunk 10]
```diff
@@ -341,6 +341,17 @@ int main(int argc, char **argv)
   source_startup_scripts(&params);
 
   /*
+   * If using the runtime (-u is not NONE),
+   * enable syntax and filetype plugins
+   */
+  if (params.use_vimrc != NULL && strcmp(params.use_vimrc, "NONE") != 0) {
```
Comment (justinmk) [inline src/nvim/main.c]: strncmp

File: src/nvim/main.c [hunk 11]
```diff
@@ -341,6 +341,17 @@ int main(int argc, char **argv)
   source_startup_scripts(&params);
 
   /*
+   * If using the runtime (-u is not NONE),
+   * enable syntax and filetype plugins
+   */
```
Comment (justinmk) [inline src/nvim/main.c]: This 4 line comment could fit on 1 line

File: runtime/doc/starting.txt (hunk 8 above)
Comment (justinmk) [inline runtime/doc/starting.txt]: Shouldn't this happen before step 4, immediately after init.Vim?

Comment (fmoralesc) [inline runtime/doc/starting.txt]: It does, the documentation is wrong (in the original version it was like documented here, but the same thing was pointed out -- I guess I forgot to update the docs). Nice catch!

File: src/nvim/main.c (hunk 11 above)
Comment (fmoralesc) [inline src/nvim/main.c]: :+1:

File: src/nvim/os/unix_defs.h (hunk 7 above)
Comment (fmoralesc) [inline src/nvim/os/unix_defs.h]: This was rebased onto master last week, I didn't see any changes related to this.

File: src/nvim/main.c (hunk 10 above)
Comment (fmoralesc) [inline src/nvim/main.c]: Is that necessary? See line 1668 and following:
~~~ c
  if (parmp->use_vimrc != NULL) {
    if (strcmp(parmp->use_vimrc, "NONE") == 0
        || strcmp(parmp->use_vimrc, "NORC") == 0) {
      if (parmp->use_vimrc[2] == 'N')
        p_lpl = false;  // don't load plugins either
    } else {
~~~

Comment (Pyrohh) [inline src/nvim/main.c]: I don't think we need to worry about it in this case: [are-char-argv-arguments-in-main-null-terminated](https://stackoverflow.com/questions/11020172/are-char-argv-arguments-in-main-null-terminated)

(here's the standard, which seems to match up: http://www.open-std.org/jtc1/sc22/wg14/www/docs/n1256.pdf)

**Comment by justinmk** [comment]
Sorry, I did not realize https://github.com/neovim/neovim/pull/3242 fell through the cracks. I merge it, so now `_PATHSEPSTR` is available in globals.h. Can you rebase on that?

**Comment by fmoralesc** [comment]
Rebased.

File: src/nvim/globals.h [hunk 12]
```diff
@@ -92,6 +92,10 @@
 # define SYNTAX_FNAME   "$VIMRUNTIME" _PATHSEPSTR "syntax" _PATHSEPSTR "%s.vim"
 #endif
 
+#ifndef SYNTAX_FILE
+# define SYNTAX_FILE   "$VIMRUNTIME" _PATHSEPSTR "syntax" _PATHSEPSTR "syntax.vim"
```
Comment (justinmk) [inline src/nvim/globals.h]: If you're not going to reuse SYNTAX_FNAME, which is nearly identical to the purpose of this, might as well hard-code this string directly in your `do_source` call in main.c. Because this isn't going to be used anywhere else--`SYNTAX_FNAME` already exists for that purpose.

File: src/nvim/globals.h [hunk 13]
```diff
@@ -92,6 +92,10 @@
...
+#endif
```
Comment (justinmk) [inline src/nvim/globals.h]: Since this is really the same as SYNTAX_FNAME, https://github.com/neovim/neovim/pull/3767 attempts to avoid the extra global.

File: src/nvim/ex_docmd.c [hunk 14]
```diff
@@ -9224,6 +9224,24 @@ static void ex_filetype(exarg_T *eap)
     EMSG2(_(e_invarg2), arg);
 }
 
+/// Same as :filetype plugin indent enable
+/// Updates the state used for :filetype without args.
+void force_enable_filetype(void)
```
Comment (justinmk) [inline src/nvim/ex_docmd.c]: "force" isn't a good name here because it actually checks if the actions were already performed and skips them if so...

**Comment by fmoralesc** [comment]
@justinmk, I have some issues with this locally (the files don't seem to be sourced anymore :/) Does it work on your end?

**Comment by justinmk** [comment]
@fmoralesc If you mean #3767 , it seems to work for me: `./build/bin/nvim -u NORC README.md` ftdetects the file as "markdown" and highlights it.

**Comment by fmoralesc** [comment]
Yeah, I couldn't make either work. It seems like there's something screwy
on my end, I had some other issues yesterday.

**Comment by justinmk** [comment]
@fmoralesc Did you ever figure out the issue on your end? I am hesitant about merging this until then.

**Comment by fmoralesc** [comment]
Ah, yes, sorry for not updating you on that! In the end, it was something with my hard drive acting up; I'm still not sure if it was hardware or software related, but certainly it had nothing to do with the code itself. I made it work by rebooting, checking the disk and freeing some space, and I haven't had those issues again.

**Comment by justinmk** [comment]
@fmoralesc Thanks. I just pushed 0.1.1, so we can move forward with this and see what everyone thinks.

**Comment by fmoralesc** [comment]
@justinmk Sounds fine. You are using the version in your PR, right?

**Comment by justinmk** [comment]
Yes

**Comment by fmoralesc** [comment]
OK, I'm closing this one then ;)
